File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Core/Bntx/BinaryData/BinaryDataReader.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.IO;
using System.Reflection;
using System.Text;
using MiniToolbox.Core.Bntx;

namespace MiniToolbox.Core.Bntx;

[DebuggerDisplay("BinaryDataReader, Position={Position}")]
public class BinaryDataReader : BinaryReader
{
	private ByteOrder _byteOrder;

	public ByteOrder ByteOrder
	{
		get
		{
			return _byteOrder;
		}
		set
		{
			_byteOrder = value;
			NeedsReversion = _byteOrder != ByteOrderHelper.SystemByteOrder;
		}
	}

	public Encoding Encoding { get; private set; }

	public bool EndOfStream => BaseStream.Position >= BaseStream.Length;

	public long Length => BaseStream.Length;

	public bool NeedsReversion { get; private set; }

	public long Position
	{
		get
		{
			return BaseStream.Position;
		}
		set
		{
			BaseStream.Position = value;
		}
	}

	public BinaryDataReader(Stream input)
		: this(input, new UTF8Encoding(), leaveOpen: false)
	{
	}

	public BinaryDataReader(Stream input, bool leaveOpen)
		: this(input, new UTF8Encoding(), leaveOpen)
	{
	}

	public BinaryDataReader(Stream input, Encoding encoding)
		: this(input, encoding, leaveOpen: false)
	{
	}

	public BinaryDataReader(Stream input, Encoding encoding, bool leaveOpen)
		: base(input, encoding, leaveOpen)
	{
		Encoding = encoding;
		ByteOrder = ByteOrderHelper.SystemByteOrder;
	}

	public void Align(int alignment)
	{
		Seek((-Position % alignment + alignment) % alignment);
	}

	public bool ReadBoolean(BinaryBooleanFormat format)
	{
		return format switch
		{
			BinaryBooleanFormat.NonZeroByte => base.ReadBoolean(), 
			BinaryBooleanFormat.NonZeroWord => ReadInt16() != 0, 
			BinaryBooleanFormat.NonZeroDword => ReadInt32() != 0, 
			_ => throw new ArgumentOutOfRangeException("format", "The specified binary boolean format is invalid."), 
		};
	}

	public bool[] ReadBooleans(int count)
	{
		return ReadMultiple(count, base.ReadBoolean);
	}

	public bool[] ReadBooleans(int count, BinaryBooleanFormat format)
	{
		bool[] array = new bool[count];
		for (int i = 0; i < array.Length; i++)
		{
			array[i] = ReadBoolean(format);
		}
		return array;
	}

	public DateTime ReadDateTime()
	{
		return ReadDateTime(BinaryDateTimeFormat.NetTicks);
	}

	public DateTime ReadDateTime(BinaryDateTimeFormat format)
	{
		return format switch
		{
			BinaryDateTimeFormat.CTime => new DateTime(1970, 1, 1).ToLocalTime().AddSeconds(ReadUInt32()), 
			BinaryDateTimeFormat.NetTicks => new DateTime(ReadInt64()), 
			_ => throw new ArgumentOutOfRangeException("format", "The specified binary date time format is invalid."), 
		};
	}

	public DateTime[] ReadDateTimes(int count)
	{
		DateTime[] array = new DateTime[count];
		for (int i = 0; i < array.Length; i++)
		{
			array[i] = ReadDateTime();
		}
		return array;
	}

	public DateTime[] ReadDateTimes(int count, BinaryDateTimeFormat format)
	{
		DateTime[] array = new DateTime[count];
		for (int i = 0; i < array.Length; i++)
		{
			array[i] = ReadDateTime(format);
		}
		return array;
	}

	public override decimal ReadDecimal()
	{
		if (NeedsReversion)
		{
			byte[] array = base.ReadBytes(16);
			Array.Reverse(array);
			return DecimalFromBytes(array);
		}
		return base.ReadDecimal();
	}

	public decimal[] ReadDecimals(int count)
	{
		return ReadMultiple(count, ReadDecimal);
	}

	public override double ReadDouble()
	{
		if (NeedsReversion)
		{
			byte[] array = base.ReadBytes(8);
			Array.Reverse(array);
			return BitConverter.ToDouble(array, 0);
		}
		return base.ReadDouble();
	}

	public double[] ReadDoubles(int count)
	{
		return ReadMultiple(count, ReadDouble);
	}

	public T ReadEnum<T>(bool strict) where T : struct, IComparable, IFormattable
	{
		return (T)ReadEnum(typeof(T), strict);
	}

	public T[] ReadEnums<T>(int count, bool strict) where T : struct, IComparable, IFormattable
	{
		T[] array = new T[count];
		for (int i = 0; i < array.Length; i++)
		{
			array[i] = ReadEnum<T>(strict);
		}
		return array;
	}

	public override short ReadInt16()
	{
		if (NeedsReversion)
		{
			byte[] array = base.ReadBytes(2);
			Array.Reverse(array);
			return BitConverter.ToInt16(array, 0);
		}
		return base.ReadInt16();
	}

	public short[] ReadInt16s(int count)
	{
		return ReadMultiple(count, ReadInt16);
	}

	public override int ReadInt32()
	{
		if (NeedsReversion)
		{
			byte[] array = base.ReadBytes(4);
			Array.Reverse(array);
			return BitConverter.ToInt32(array, 0);
		}
		return base.ReadInt32();
	}

	public int[] ReadInt32s(int count)
	{
		return ReadMultiple(count, ReadInt32);
	}

	public override long ReadInt64()
	{
		if (NeedsReversion)
		{
			byte[] array = base.ReadBytes(8);
			Array.Reverse(array);
			return BitConverter.ToInt64(array, 0);
		}
		return base.ReadInt64();
	}

	public long[] ReadInt64s(int count)
	{
		return ReadMultiple(count, ReadInt64);
	}

	public T ReadObject<T>()
	{
		return (T)ReadObject(null, BinaryMemberAttribute.Default, typeof(T));
	}

	public T[] ReadObjects<T>(int count)
	{
		return ReadMultiple(count, ReadObject<T>);
	}

	public sbyte[] ReadSBytes(int count)
	{
		return ReadMultiple(count, ReadSByte);
	}

	public override float ReadSingle()
	{
		if (NeedsReversion)
		{
			byte[] array = base.ReadBytes(4);
			Array.Reverse(array);
			return BitConverter.ToSingle(array, 0);
		}
		return base.ReadSingle();
	}

	public float[] ReadSingles(int count)
	{
		return ReadMultiple(count, ReadSingle);
	}

	public string ReadString(BinaryStringFormat format)
	{
		return ReadString(format, Encoding);
	}

	public string ReadString(BinaryStringFormat format, Encoding encoding)
	{
		return format switch
		{
			BinaryStringFormat.ByteLengthPrefix => ReadStringInternal(ReadByte(), encoding), 
			BinaryStringFormat.WordLengthPrefix => ReadStringInternal(ReadInt16(), encoding), 
			BinaryStringFormat.DwordLengthPrefix => ReadStringInternal(ReadInt32(), encoding), 
			BinaryStringFormat.VariableLengthPrefix => ReadStringInternal(Read7BitEncodedInt(), encoding), 
			BinaryStringFormat.ZeroTerminated => ReadZeroTerminatedString(encoding), 
			BinaryStringFormat.NoPrefixOrTermination => throw new ArgumentException("NoPrefixOrTermination cannot be used for read operations if no length has been specified.", "format"), 
			_ => throw new ArgumentOutOfRangeException("format", "The specified binary string format is invalid."), 
		};
	}

	public string ReadString(int length)
	{
		return ReadString(length, Encoding);
	}

	public string ReadString(int length, Encoding encoding)
	{
		return encoding.GetString(ReadBytes(length));
	}

	public string[] ReadStrings(int count)
	{
		string[] array = new string[count];
		for (int i = 0; i < array.Length; i++)
		{
			array[i] = ReadString();
		}
		return array;
	}

	public string[] ReadStrings(int count, BinaryStringFormat format)
	{
		string[] array = new string[count];
		for (int i = 0; i < array.Length; i++)
		{
			array[i] = ReadString(format);
		}
		return array;
	}

	public string[] ReadStrings(int count, BinaryStringFormat format, Encoding encoding)
	{
		string[] array = new string[count];
		for (int i = 0; i < array.Length; i++)
		{
			array[i] = ReadString(format, encoding);
		}
		return array;
	}

	public string[] ReadStrings(int count, int length)
	{
		string[] array = new string[count];
		for (int i = 0; i < array.Length; i++)
		{
			array[i] = ReadString(length);
		}
		return array;
	}

	public string[] ReadStrings(int count, int length, Encoding encoding)
	{
		string[] array = new string[count];
		for (int i = 0; i < array.Length; i++)
		{
			array[i] = ReadString(length, Encoding);
		}
		return array;
	}

	public override ushort ReadUInt16()
	{
		if (NeedsReversion)
		{
			byte[] array = base.ReadBytes(2);
			Array.Reverse(array);
			return BitConverter.ToUInt16(array, 0);
		}
		return base.ReadUInt16();
	}

	public ushort[] ReadUInt16s(int count)
	{
		return ReadMultiple(count, ReadUInt16);
	}

	public override uint ReadUInt32()
	{
		if (NeedsReversion)
		{
			byte[] array = base.ReadBytes(4);
			Array.Reverse(array);
			return BitConverter.ToUInt32(array, 0);
		}
		return base.ReadUInt32();
	}

	public uint[] ReadUInt32s(int count)
	{
		return ReadMultiple(count, ReadUInt32);
	}

	public override ulong ReadUInt64()
	{
		if (NeedsReversion)
		{
			byte[] array = base.ReadBytes(8);
			Array.Reverse(array);
			return BitConverter.ToUInt64(array, 0);
		}
		return base.ReadUInt64();
	}

	public ulong[] ReadUInt64s(int count)
	{
		return ReadMultiple(count, ReadUInt64);
	}

	public long Seek(long offset)
	{
		return Seek(offset, SeekOrigin.Current);
	}

	public long Seek(long offset, SeekOrigin origin)
	{
		return BaseStream.Seek(offset, origin);
	}

	public SeekTask TemporarySeek()
	{
		return TemporarySeek(0L, SeekOrigin.Current);
	}

	public SeekTask TemporarySeek(long offset)
	{
		return TemporarySeek(offset, SeekOrigin.Current);
	}

	public SeekTask TemporarySeek(long offset, SeekOrigin origin)
	{
		return new SeekTask(BaseStream, offset, origin);
	}

	private T[] ReadMultiple<T>(int count, Func<T> readFunc)
	{
		T[] array = new T[count];
		for (int i = 0; i < count; i++)
		{
			array[i] = readFunc();
		}
		return array;
	}

	private decimal DecimalFromBytes(byte[] bytes)
	{
		if (bytes.Length < 16)
		{
			throw new ArgumentException("Not enough bytes to convert decimal from.", "bytes");
		}
		int[] array = new int[4];
		for (int i = 0; i < 16; i += 4)
		{
			array[i / 4] = BitConverter.ToInt32(bytes, i);
		}
		return new decimal(array);
	}

	private object ReadEnum(Type type, bool strict)
	{
		object obj = ReadObject(null, BinaryMemberAttribute.Default, Enum.GetUnderlyingType(type));
		if (strict && !EnumExtensions.IsValid(type, obj))
		{
			throw new InvalidDataException($"Read value {obj} is not defined in the given enum type {type}.");
		}
		return obj;
	}

	private object ReadObject(object instance, BinaryMemberAttribute attribute, Type type)
	{
		if (attribute.Converter == null)
		{
			if (type == typeof(string))
			{
				if (attribute.StringFormat == BinaryStringFormat.NoPrefixOrTermination)
				{
					return ReadString(attribute.Length);
				}
				return ReadString(attribute.StringFormat);
			}
			if (type.IsEnumerable())
			{
				throw new InvalidOperationException("Multidimensional arrays cannot be read directly.");
			}
			if (type == typeof(bool))
			{
				return ReadBoolean(attribute.BooleanFormat);
			}
			if (type == typeof(byte))
			{
				return ReadByte();
			}
			if (type == typeof(DateTime))
			{
				return ReadDateTime(attribute.DateTimeFormat);
			}
			if (type == typeof(decimal))
			{
				return ReadDecimal();
			}
			if (type == typeof(double))
			{
				return ReadDouble();
			}
			if (type == typeof(short))
			{
				return ReadInt16();
			}
			if (type == typeof(int))
			{
				return ReadInt32();
			}
			if (type == typeof(long))
			{
				return ReadInt64();
			}
			if (type == typeof(sbyte))
			{
				return ReadSByte();
			}
			if (type == typeof(float))
			{
				return ReadSingle();
			}
			if (type == typeof(ushort))
			{
				return ReadUInt16();
			}
			if (type == typeof(uint))
			{
				return ReadUInt32();
			}
			if (type == typeof(ulong))
			{
				return ReadUInt64();
			}
			if (type.GetTypeInfo().IsEnum)
			{
				return ReadEnum(type, attribute.Strict);
			}
			return ReadCustomObject(type, null, Position);
		}
		return BinaryConverterCache.GetConverter(attribute.Converter).Read(this, instance, attribute);
	}

	private object ReadCustomObject(Type type, object instance, long startOffset)
	{
		TypeData typeData = TypeData.GetTypeData(type);
		instance = instance ?? typeData.GetInstance();
		if (typeData.Attribute.Inherit && typeData.TypeInfo.BaseType != null)
		{
			ReadCustomObject(typeData.TypeInfo.BaseType, instance, startOffset);
		}
		foreach (MemberData member in typeData.Members)
		{
			if (member.Attribute.OffsetOrigin == OffsetOrigin.Begin)
			{
				Position = startOffset + member.Attribute.Offset;
			}
			else if (member.Attribute.Offset != 0)
			{
				Position += member.Attribute.Offset;
			}
			Type enumerableElementType = member.Type.GetEnumerableElementType();
			object value;
			if (enumerableElementType == null)
			{
				value = ReadObject(instance, member.Attribute, member.Type);
			}
			else
			{
				Array array = Array.CreateInstance(enumerableElementType, member.Attribute.Length);
				for (int i = 0; i < array.Length; i++)
				{
					array.SetValue(ReadObject(instance, member.Attribute, enumerableElementType), i);
				}
				value = array;
			}
			MemberInfo memberInfo = member.MemberInfo;
			if ((object)memberInfo == null)
			{
				continue;
			}
			if (!(memberInfo is FieldInfo fieldInfo))
			{
				if (memberInfo is PropertyInfo propertyInfo)
				{
					propertyInfo.SetValue(instance, value);
				}
			}
			else
			{
				fieldInfo.SetValue(instance, value);
			}
		}
		return instance;
	}

	private string ReadStringInternal(int length, Encoding encoding)
	{
		return encoding.GetString(ReadBytes(length * encoding.GetByteCount("a")));
	}

	private string ReadZeroTerminatedString(Encoding encoding)
	{
		int byteCount = encoding.GetByteCount("a");
		List<byte> list = new List<byte>();
		switch (byteCount)
		{
		case 1:
		{
			for (byte b = ReadByte(); b != 0; b = ReadByte())
			{
				list.Add(b);
			}
			break;
		}
		case 2:
		{
			for (uint num = ReadUInt16(); num != 0; num = ReadUInt16())
			{
				byte[] bytes = BitConverter.GetBytes(num);
				list.Add(bytes[0]);
				list.Add(bytes[1]);
			}
			break;
		}
		}
		return encoding.GetString(list.ToArray());
	}
}

File: src/Starfield2026.MiniToolbox/src/MiniToolbox.Api/src/bntx/BinaryData/BinaryDataReader.ts
import { ByteOrder } from './ByteOrder.js';
import { ByteOrderHelper } from './ByteOrderHelper.js';
import { BinaryBooleanFormat } from './BinaryBooleanFormat.js';
import { BinaryDateTimeFormat } from './BinaryDateTimeFormat.js';
import { BinaryStringFormat } from './BinaryStringFormat.js';
import { BinaryMemberAttribute } from './BinaryMemberAttribute.js';
import { BinaryConverterCache } from './BinaryConverterCache.js';
import { TypeData } from './TypeData.js';
import { OffsetOrigin } from './OffsetOrigin.js';
import { EnumExtensions } from './EnumExtensions.js';
import { SeekTask } from './SeekTask.js';

/**
 * 1:1 port of BinaryDataReader.cs
 * Full-featured binary reader with endianness support, object deserialization,
 * and all the string/boolean/datetime format variants.
 */
export class BinaryDataReader {
    private _buffer: Buffer;
    private _position: number = 0;
    private _byteOrder: ByteOrder;
    private _needsReversion: boolean = false;

    get byteOrder(): ByteOrder { return this._byteOrder; }
    set byteOrder(value: ByteOrder) {
        this._byteOrder = value;
        this._needsReversion = this._byteOrder !== ByteOrderHelper.systemByteOrder;
    }

    get needsReversion(): boolean { return this._needsReversion; }
    get endOfStream(): boolean { return this._position >= this._buffer.length; }
    get length(): number { return this._buffer.length; }

    get position(): number { return this._position; }
    set position(value: number) { this._position = value; }

    get buffer(): Buffer { return this._buffer; }

    constructor(buffer: Buffer, byteOrder?: ByteOrder) {
        this._buffer = buffer;
        this._byteOrder = byteOrder ?? ByteOrderHelper.systemByteOrder;
        this._needsReversion = this._byteOrder !== ByteOrderHelper.systemByteOrder;
    }

    // ── Alignment ──

    align(alignment: number): void {
        this.seek((-this._position % alignment + alignment) % alignment);
    }

    // ── Boolean ──

    readBoolean(format?: BinaryBooleanFormat): boolean {
        const fmt = format ?? BinaryBooleanFormat.NonZeroByte;
        switch (fmt) {
            case BinaryBooleanFormat.NonZeroByte: return this.readByte() !== 0;
            case BinaryBooleanFormat.NonZeroWord: return this.readInt16() !== 0;
            case BinaryBooleanFormat.NonZeroDword: return this.readInt32() !== 0;
            default: throw new Error('The specified binary boolean format is invalid.');
        }
    }

    readBooleans(count: number, format?: BinaryBooleanFormat): boolean[] {
        return this.readMultiple(count, () => this.readBoolean(format));
    }

    // ── DateTime ──

    readDateTime(format?: BinaryDateTimeFormat): Date {
        const fmt = format ?? BinaryDateTimeFormat.NetTicks;
        switch (fmt) {
            case BinaryDateTimeFormat.CTime: {
                const seconds = this.readUInt32();
                return new Date(seconds * 1000);
            }
            case BinaryDateTimeFormat.NetTicks: {
                const ticks = this.readInt64();
                // .NET ticks are 100-nanosecond intervals since 0001-01-01
                const epochTicks = BigInt('621355968000000000'); // Ticks to Unix epoch
                const msSinceEpoch = Number((BigInt(ticks) - epochTicks) / BigInt(10000));
                return new Date(msSinceEpoch);
            }
            default: throw new Error('The specified binary date time format is invalid.');
        }
    }

    readDateTimes(count: number, format?: BinaryDateTimeFormat): Date[] {
        return this.readMultiple(count, () => this.readDateTime(format));
    }

    // ── Primitives (with endian swap) ──

    readByte(): number {
        const val = this._buffer.readUInt8(this._position);
        this._position += 1;
        return val;
    }

    readBytes(count: number): Buffer {
        const slice = this._buffer.subarray(this._position, this._position + count);
        this._position += count;
        return slice;
    }

    readSByte(): number {
        const val = this._buffer.readInt8(this._position);
        this._position += 1;
        return val;
    }

    readSBytes(count: number): number[] {
        return this.readMultiple(count, () => this.readSByte());
    }

    readInt16(): number {
        const val = this._needsReversion
            ? this._buffer.readInt16BE(this._position)
            : this._buffer.readInt16LE(this._position);
        this._position += 2;
        return val;
    }

    readInt16s(count: number): number[] {
        return this.readMultiple(count, () => this.readInt16());
    }

    readUInt16(): number {
        const val = this._needsReversion
            ? this._buffer.readUInt16BE(this._position)
            : this._buffer.readUInt16LE(this._position);
        this._position += 2;
        return val;
    }

    readUInt16s(count: number): number[] {
        return this.readMultiple(count, () => this.readUInt16());
    }

    readInt32(): number {
        const val = this._needsReversion
            ? this._buffer.readInt32BE(this._position)
            : this._buffer.readInt32LE(this._position);
        this._position += 4;
        return val;
    }

    readInt32s(count: number): number[] {
        return this.readMultiple(count, () => this.readInt32());
    }

    readUInt32(): number {
        const val = this._needsReversion
            ? this._buffer.readUInt32BE(this._position)
            : this._buffer.readUInt32LE(this._position);
        this._position += 4;
        return val;
    }

    readUInt32s(count: number): number[] {
        return this.readMultiple(count, () => this.readUInt32());
    }

    readInt64(): number {
        const val = this._needsReversion
            ? Number(this._buffer.readBigInt64BE(this._position))
            : Number(this._buffer.readBigInt64LE(this._position));
        this._position += 8;
        return val;
    }

    readInt64s(count: number): number[] {
        return this.readMultiple(count, () => this.readInt64());
    }

    readUInt64(): number {
        const val = this._needsReversion
            ? Number(this._buffer.readBigUInt64BE(this._position))
            : Number(this._buffer.readBigUInt64LE(this._position));
        this._position += 8;
        return val;
    }

    readUInt64s(count: number): number[] {
        return this.readMultiple(count, () => this.readUInt64());
    }

    readSingle(): number {
        const val = this._needsReversion
            ? this._buffer.readFloatBE(this._position)
            : this._buffer.readFloatLE(this._position);
        this._position += 4;
        return val;
    }

    readSingles(count: number): number[] {
        return this.readMultiple(count, () => this.readSingle());
    }

    readDouble(): number {
        const val = this._needsReversion
            ? this._buffer.readDoubleBE(this._position)
            : this._buffer.readDoubleLE(this._position);
        this._position += 8;
        return val;
    }

    readDoubles(count: number): number[] {
        return this.readMultiple(count, () => this.readDouble());
    }

    readDecimal(): number {
        // .NET decimal is 128 bits: 4 int32s. bits[0..2] = 96-bit mantissa, bits[3] = sign + scale.
        // Port of C# DecimalFromBytes: new decimal(int[4])
        const bytes = this.readBytes(16);
        let buf = bytes;
        if (this._needsReversion) {
            buf = Buffer.from(bytes).reverse();
        }
        const lo = buf.readInt32LE(0);
        const mid = buf.readInt32LE(4);
        const hi = buf.readInt32LE(8);
        const flags = buf.readInt32LE(12);

        const negative = (flags & 0x80000000) !== 0;
        const scale = (flags >>> 16) & 0xFF;

        // Reconstruct 96-bit mantissa as a JS number (may lose precision beyond 2^53)
        const mantissa = ((hi >>> 0) * 0x100000000 + (mid >>> 0)) * 0x100000000 + (lo >>> 0);
        const value = mantissa / Math.pow(10, scale);
        return negative ? -value : value;
    }

    readDecimals(count: number): number[] {
        return this.readMultiple(count, () => this.readDecimal());
    }

    // ── String ──

    readString(formatOrLength?: BinaryStringFormat | number): string {
        if (typeof formatOrLength === 'number') {
            return this.readStringWithLength(formatOrLength);
        }
        const format: BinaryStringFormat = formatOrLength ?? BinaryStringFormat.VariableLengthPrefix;
        switch (format as BinaryStringFormat) {
            case BinaryStringFormat.ByteLengthPrefix:
                return this.readStringInternal(this.readByte());
            case BinaryStringFormat.WordLengthPrefix:
                return this.readStringInternal(this.readInt16());
            case BinaryStringFormat.DwordLengthPrefix:
                return this.readStringInternal(this.readInt32());
            case BinaryStringFormat.VariableLengthPrefix:
                return this.readStringInternal(this.read7BitEncodedInt());
            case BinaryStringFormat.ZeroTerminated:
                return this.readZeroTerminatedString();
            case BinaryStringFormat.NoPrefixOrTermination:
                throw new Error('NoPrefixOrTermination cannot be used for read operations if no length has been specified.');
            default:
                throw new Error('The specified binary string format is invalid.');
        }
    }

    readStringWithLength(length: number): string {
        const bytes = this.readBytes(length);
        return bytes.toString('utf8');
    }

    readStrings(count: number, formatOrLength?: BinaryStringFormat | number): string[] {
        return this.readMultiple(count, () => this.readString(formatOrLength));
    }

    // ── Enum ──

    readEnum<T extends number>(enumObj: Record<string, number | string>, strict: boolean = false): T {
        // Read the underlying value (assumed to be the same width as the enum's underlying type)
        // For simplicity, read as int32 — callers should use readByte/readUInt16/etc. for smaller enums
        const value = this.readInt32();
        if (strict && !EnumExtensions.isValid(enumObj, value)) {
            throw new Error(`Read value ${value} is not defined in the given enum type.`);
        }
        return value as T;
    }

    // ── Object (reflection-based deserialization) ──

    readObject(typeName: string): any {
        return this.readObjectInternal(null, BinaryMemberAttribute.Default, typeName);
    }

    readObjects(count: number, typeName: string): any[] {
        return this.readMultiple(count, () => this.readObject(typeName));
    }

    private readObjectInternal(instance: any, attribute: BinaryMemberAttribute, typeName: string): any {
        if (attribute.converter !== null) {
            const converter = BinaryConverterCache.getConverter(attribute.converter);
            return converter.read(this, instance, attribute);
        }

        // Primitive type dispatch (matching C# ReadObject)
        switch (typeName) {
            case 'string':
                if (attribute.stringFormat === BinaryStringFormat.NoPrefixOrTermination) {
                    return this.readStringWithLength(attribute.length);
                }
                return this.readString(attribute.stringFormat);
            case 'boolean': return this.readBoolean(attribute.booleanFormat);
            case 'byte': return this.readByte();
            case 'sbyte': return this.readSByte();
            case 'int16': return this.readInt16();
            case 'uint16': return this.readUInt16();
            case 'int32': return this.readInt32();
            case 'uint32': return this.readUInt32();
            case 'int64': return this.readInt64();
            case 'uint64': return this.readUInt64();
            case 'float': return this.readSingle();
            case 'double': return this.readDouble();
            case 'decimal': return this.readDecimal();
            case 'datetime': return this.readDateTime(attribute.dateTimeFormat);
            default:
                // Custom object — use TypeData registry
                return this.readCustomObject(typeName, instance, this._position);
        }
    }

    private readCustomObject(typeName: string, instance: any, startOffset: number): any {
        const typeData = TypeData.getTypeData(typeName);
        instance = instance ?? typeData.getInstance();

        for (const member of typeData.members) {
            if (member.attribute.offsetOrigin === OffsetOrigin.Begin) {
                this._position = startOffset + member.attribute.offset;
            } else if (member.attribute.offset !== 0) {
                this._position += member.attribute.offset;
            }

            let value: any;
            if (member.attribute.length > 0 && member.type.endsWith('[]')) {
                // Array member
                const elementType = member.type.replace('[]', '');
                const arr: any[] = [];
                for (let i = 0; i < member.attribute.length; i++) {
                    arr.push(this.readObjectInternal(instance, member.attribute, elementType));
                }
                value = arr;
            } else {
                value = this.readObjectInternal(instance, member.attribute, member.type);
            }

            member.setValue(instance, value);
        }

        return instance;
    }

    // ── Seek ──

    seek(offset: number, origin?: 'begin' | 'current' | 'end'): number {
        const o = origin ?? 'current';
        switch (o) {
            case 'begin': this._position = offset; break;
            case 'current': this._position += offset; break;
            case 'end': this._position = this._buffer.length + offset; break;
        }
        return this._position;
    }

    temporarySeek(offset?: number, origin?: 'begin' | 'current' | 'end'): SeekTask {
        return new SeekTask(this, offset ?? 0, origin ?? 'current');
    }

    // ── Private helpers ──

    private readMultiple<T>(count: number, readFunc: () => T): T[] {
        const arr = new Array<T>(count);
        for (let i = 0; i < count; i++) {
            arr[i] = readFunc();
        }
        return arr;
    }

    private readStringInternal(length: number): string {
        const bytes = this.readBytes(length);
        return bytes.toString('utf8');
    }

    private readZeroTerminatedString(): string {
        const bytes: number[] = [];
        let b = this.readByte();
        while (b !== 0) {
            bytes.push(b);
            if (this.endOfStream) break;
            b = this.readByte();
        }
        return Buffer.from(bytes).toString('utf8');
    }

    private read7BitEncodedInt(): number {
        let result = 0;
        let shift = 0;
        let b: number;
        do {
            b = this.readByte();
            result |= (b & 0x7F) << shift;
            shift += 7;
        } while (b >= 0x80);
        return result;
    }
}
